src/rust/src/encoder/common.rs

```diff
@@ -0,0 +1,483 @@
+                            dest[dest_idx] = b'?';
+                        }
+                        dest_idx += 1;
+                    } else {
+                        dest[dest_idx] = b'?';
+                        dest_idx += 1;
+                    }
+                } else {
+                    dest[dest_idx] = b'?';
+                    dest_idx += 1;
+                }
+            }
+            Encoding::Ucs2 => {
+                return EncoderError::Unsupported as i32;
+            }
+            Encoding::Line21 => {
+                if c_len == 1 {
+                    dest[dest_idx] = src[src_idx];
+                    dest_idx += 1;
+                } else {
+                    dest[dest_idx] = b'?';
+                    dest_idx += 1;
+                }
+            }
+        }
+        src_idx += c_len;
+    }
+
+    if dest_idx < dest.len() {
+        dest[dest_idx] = 0;
+    }
+    dest_idx as i32
+}
+#[allow(unused_variables)]
+fn change_latin1_encoding(dest: &mut [u8], src: &[u8], len: i32, out_enc: Encoding) -> i32 {
+    EncoderError::Unsupported as i32
+}
+
+#[allow(unused_variables)]
+fn change_unicode_encoding(dest: &mut [u8], src: &[u8], len: i32, out_enc: Encoding) -> i32 {
+    EncoderError::Unsupported as i32
+}
+
+pub fn change_ascii_encoding(
+    dest: &mut Vec<u8>,
+    src: &[u8],
+    out_enc: Encoding,
+) -> Result<usize, i32> {
+    dest.clear();
+
+    for &c in src {
+        match out_enc {
+            Encoding::Utf8 => {
+                let utf8_packed = line21_to_utf8(c);
+                let bytes = if utf8_packed <= 0xff {
+                    vec![utf8_packed as u8]
+                } else if utf8_packed <= 0xffff {
+                    vec![(utf8_packed >> 8) as u8, utf8_packed as u8]
+                } else if utf8_packed <= 0xffffff {
+                    vec![
+                        (utf8_packed >> 16) as u8,
+                        (utf8_packed >> 8) as u8,
+                        utf8_packed as u8,
+                    ]
+                } else {
+                    return Err(-1); // Invalid byte count
+                };
+                dest.extend(bytes);
+            }
+            Encoding::Latin1 => {
+                let latin1_char = line21_to_latin1(c);
+                dest.push(latin1_char);
+            }
+            Encoding::Ucs2 => {
+                let ucs2_char = line21_to_ucs2(c);
+                // UCS-2 is 2 bytes, little-endian
+                dest.extend_from_slice(&ucs2_char.to_le_bytes());
+            }
+            Encoding::Line21 => {
+                dest.extend_from_slice(src);
+                return Ok(src.len());
+            }
+        }
+    }
+
+    // Add null terminator
+    dest.push(0);
+
+    Ok(dest.len() - 1) // Return length without null terminator
+}
+
+fn utf8_to_latin1_map(character: u32) -> Latin1Char {
+    ucs2_to_latin1(char_to_ucs2(char::try_from(character).unwrap()))
+}
+pub fn get_str_basic(
```

Comment:
Should we keep pure rust function in rust/lib_ccxr?

I think lib_ccxr was meant to keep pure rust implementation of lib_ccx and the top level rust for interoperability between lib_ccxr and lib_ccx.